Add documentation on screenshots in pubspec.md (#4259)
dart-lang/site-www: src/tools/pub/pubspec.md

: Optional. Specify files to ignore when conducting a pre-publishing search
  for potential leaks of secrets.
  [_Learn more._](#false_secrets)
  
`screenshots`

> : Optional. Specify a list of screenshot files to be shown 

Reviewer: ```suggestion
: Optional. Specify a list of screenshot files to display
```

Author: Done.

{{site.alert.end}}

{{site.alert.version-note}}
  Dart 2.15 added support for the `false_secrets` field.
{{site.alert.end}}

### Screenshots

Packages can showcase their widgets or other visual elements using screenshots

> displayed on their package page.

Reviewer: **nit:** Don't need to say "package page". The context is sufficient.
```suggestion
displayed on their page.
```

Author: Done.

A package can list up to 10 screenshots under the `screenshots` field.
Each screenshot includes one `description` and one `path`. 
The `description` explains what the screenshot depicts in no more than 160 
characters. 
For example:

```yaml
screenshots:
  - description: 'This screenshot shows the transformation of a number of bytes 
  to a human-readable expression.'
    path: path/to/image/in/package/500x500.webp
  - description: 'This screenshot shows a stack trace returning a human-readable
  representation.'
    path: path/to/image/in/package.png
```

> Screenshots are limited to:

Reviewer: **nit:** Active voice.
```suggestion
Pub.dev limits screenshots to the following specifications:
```

Author: done.

> - File size: max 4 MB per image.

Reviewer 2: ```suggestion

- File size: max 4 MB per image.
```

Author: Done.

- File types: `png`, `jpg`, `gif`, or `webp`. 
- Static and animated images are both allowed.

> It is recommended to keep screenshot files small, since the files will be 
> included in each download of the package.

Reviewer: **suggestion:** We can express our opinion here as the way to go. No need to couch your language.
```suggestion
Keep screenshot files small. Each download of the package includes all screenshot files.
```

Author: Done.

  Support for the `false_secrets` field was added in Dart 2.15.
{{site.alert.end}}

### Screenshots

Packages may showcase their widgets or other visual elements using screenshots.

A package can include up to 10 screenshots. These are specified in the pubspec as a list under the `screenshots` field. 
Each screenshot is specified by a `description` and a `path`. The `description` is a string of at most 160 characters describing the screenshot. 
For example:

```yaml
screenshots:
  - description: 'description of screenshot'
    path: path/to/image/in/package/500x500.webp
  - description: 'description of second screenshot'
    path: path/to/image/in/package.png
```

Supported file types include : `png`, `jpg`, `gif`, `webp`. Both static and animated files are supported. The screenshots shown on the pub.dev site
will all be `webp` images generated from the original image. 

> The first screenshot in the list will be used for a thumbnail. Thumbnails will always be static. In case the original image is animated the first frame will be used for the thumbnail. We generate quadratic thumbnails to be shown on the pub.dev site. Hence, for the first screenshot, we recommend a 1:1 aspect ratio for best fitting. If the ratio is not 1:1, the image will be resized keeping the original aspect ratio. 

Reviewer 3: ```suggestion
The first screenshot in the list will be used for a thumbnail. Thumbnails will always be static. In case the original image is animated the first frame will be used for the thumbnail. Pub.dev generates square thumbnails, so, for the first screenshot, we recommend a 1:1 aspect ratio. If the ratio is not 1:1, the image will be resized to fit in a square, but keeping the original aspect ratio. 
```

> The maximum allowed image file size is 4MB.

Reviewer 3: Should we add a reminder that the files also will be included in each download of the package so it is preferred to keep them small?

Author: Hmm, I think that is quite clear, but maybe I am biased :-)

Reviewer 3: Not saying it's unclear - just a friendly reminder about the implications of including large screenshots.

Reviewer 2: 4MB seems pretty large for a screenshot of a widget. I think it wouldn't hurt to have a small mention about this, since I imagine many users don't consider how a package download includes everything.

Do these images contribute to the 100 MB cap after gzip compression? That may make this mention even more relevant.

Author: Added a note.

>   Support for the `false_secrets` field was added in Dart 2.15.

Reviewer: **issue:** Avoid passive voice.
```suggestion
  Dart 2.15 added support for the `false_secrets` field.
```

Author: Done.

{{site.alert.end}}

### Screenshots

> Packages may showcase their widgets or other visual elements using screenshots.

Reviewer 2: Should we explain where these screenshots will be showcased as this is the section opening? Perhaps, something like:

```suggestion
Packages may showcase their widgets or other visual elements using screenshots
that are displayed on their package page.
```

Reviewer 4: Maybe, the verb is  "can".

Because it's something you control as package author. And these docs are mostly geared towards package authors.

Reviewer 2: You're right, edited the suggestion. Thanks.

Author: Done.

> A package can include up to 10 screenshots. 
> These are specified in the pubspec as a list under the `screenshots` field. 

Reviewer 2: Perhaps we can simplify this mentioning, relying on the context of this being the pubspec page?

```suggestion
A package can specify up to 10 screenshots
as a list under the `screenshots` field.
```

Author: Done.

> Each screenshot is specified by a `description` and a `path`. 
> The `description` is a string of at most 160 characters
> describing the screenshot. 

Reviewer: **issue:** Revise to remove passive voice and the circular description.
```suggestion
Specify each screenshot with a `description` and a `path`. 
For the `description`, provide a string of no more than 160 characters
that explains what the screenshot depicts. 
```

Reviewer 4: I'm by no means an expert on writing good documentation, text and such... :rofl: 
But isn't it reasonably fair to say that `https://dart.dev/tools/pub/pubspec` is _reference style_ documentation?

It does not outline how you can publish a package or what you should/can do.
But it aims to document how properties in `pubspec.yaml` are interpreted by pub / pub.dev, what they mean. Whether you're looking to write packages or consume packages. My argument is that this is the closest we have to reference style documentation of the `pubspec.yaml` file format.

I think that's also reflected in the sitemap / menu layout:
![image](https://user-images.githubusercontent.com/149732/197514955-4f26fd64-cc78-44bd-84b4-846e23c87899.png)

For documentation on what you as a package author should do. We have "guide"-style documentation in:
https://dart.dev/tools/pub/publishing
(which should also be updated with guidelines for how to use screenshots).

---------------------------------

So maybe, we don't want to write:
> For the `description`, provide a string of no more than 160 characters
> that explains what the screenshot depicts. 

But use a more passive voice?

https://developers.google.com/style/reference-verbs

Reviewer: @jonasfj : Great points actually. Maybe more along the lines of:
```
Each screenshot includes one `description` and one `path`. 
The `description` explains what the screenshot depicts in no more than 160 characters. 
```

Author: Done, using the last suggestion.

For example:

```yaml
screenshots:

>   - description: 'description of screenshot'
>     path: path/to/image/in/package/500x500.webp
>   - description: 'description of second screenshot'
>     path: path/to/image/in/package.png

Reviewer: **issue:** Avoid a circular definition.
```suggestion
  - description: 'This screenshot shows the transformation of a number of bytes to a human-readable expression.'
    path: path/to/image/in/package/500x500.webp
  - description: 'This screenshot shows a stack trace returning a human-readable representation.'
    path: path/to/image/in/package.png
```

Author: Done.

```

> 

Reviewer: **suggestion:** Change the following specifications to a bulleted list.

Example:
```
Screenshots must meet the following requirements:
- Use one of the following file types: `png`, `jpg`, `gif`, and `webp`. 
- Use a static or animated image.
- Keep to a file size of less than 4 MB each.
- Use a static, square image for your first screenshot. 
  Pub.dev generates the package's thumbnail image from the first screenshot.
  - If this screenshot uses animation, pub.dev uses its first frame.
  - If this screenshot is not square, pub.dev (stretches|trims|pads) the thumbnail.
```

> Supported file types include: `png`, `jpg`, `gif`, and `webp`. 
> Both static and animated files are supported. 

Reviewer: **issue:** Avoid passive voice.
```suggestion
This supports static and animated versions of the following file types: `png`, `jpg`, `gif`, and `webp`. 
```

The screenshots shown on the pub.dev site

> will all be `webp` images generated from the original image. 

Author: Done, removed.

> The first screenshot in the list will be used for a thumbnail. 

Reviewer 2: ```suggestion
The first screenshot in the list
will be used as the package's thumbnail. 
```

Author: Done.

Thumbnails will always be static. 

> In case the original image is animated,

Reviewer 2: ```suggestion
If the original image is animated,
```

Author: Done.

the first frame will be used for the thumbnail. 
Pub.dev generates square thumbnails, so, 
for the first screenshot, 
we recommend a 1:1 aspect ratio. 
If the ratio is not 1:1, 
the image will be resized to fit in a square, 

> but keeping the original aspect ratio. 

Reviewer 2: What does this mean? Does it add margins to the image?

Author: It doesn't add margins. The image is scaled down such that the largest dimension is reduced to the size of the square. And then the other dimension is resized according to the aspect ratio. (The image will appear in the center of the square.) Does it make sense? :-)

Author: In any case I removed this section. See my comment below.
